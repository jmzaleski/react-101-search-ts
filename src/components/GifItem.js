import React from 'react';

// no state or props. just a wrapper function around a gif and associated callback
const GifItem = ({gif, onGifSelect}) => {

    return (
        <div className="gif-item" onClick={() => onGifSelect(gif)}>
            <img src={gif.images.downsized.url} alt={'foo'} />
        </div>
    )
};

export default GifItem;