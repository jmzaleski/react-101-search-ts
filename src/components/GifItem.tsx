import * as React from 'react';

// temporary experimenting with contributors' .d.ts file
// See https://github.com/rawrmaan/restyped-giphy-api

import * as giphy from 'restyped-giphy-api';

type tOnGifSelectCallback = (gifobj: giphy.GIFObject) => any;

// The GifItem component takes two props from its parent
// 1. GIFObject as retured from the giphy search endpoint
// 2. a function to call back when the gifitem is clicked

export interface IGifItemProps {
    gifobj: giphy.GIFObject,  
    onGifSelectCallback: tOnGifSelectCallback,
    key: string
    };

// TODO: what about that key prop? it's just floating around in GifItem. stuff it in alt.

export function GifItem(props: IGifItemProps) {
    return (
// tslint:disable-next-line jsx-no-lambda 
        <div className="gif-item" onClick={() => props.onGifSelectCallback(props.gifobj)}>
            <img src={props.gifobj.images.downsized.url} alt={props.key} />
        </div>
    )
};
