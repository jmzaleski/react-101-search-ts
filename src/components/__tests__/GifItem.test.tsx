import * as enzyme from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import * as React from 'react';
import * as ReactShallowRenderer from 'react-test-renderer/shallow';
import * as giphy from 'restyped-giphy-api';
import { GifItem, tOnGifSelectCallback } from '../GifItem';

// mock of a giphy.GIFObject
const MockedConstructor = jest.fn<giphy.GIFObject>(() => {
    return { images: { downsized: { url: 'mockurl' } } }
});

// JSX translated GifItem
let ele: any;
let mockGiphyGifObject: giphy.GIFObject;
let mockedCallbackForGifItem: jest.Mock<tOnGifSelectCallback>;

beforeEach(() => {
    mockGiphyGifObject = MockedConstructor();
    mockedCallbackForGifItem = jest.fn((gifobj:giphy.GIFObject) => null);
    ele = <GifItem gifobj={mockGiphyGifObject} onGifSelectCallback={mockedCallbackForGifItem} key={"foo"}/> ;
});
    

// shallow render to snapshot file
// turns out the file is useful to snoop at to figure out selector (see next test)
describe('<GiphItem />', () => {
    it('renders', () => {
      expect(ReactShallowRenderer.createRenderer().render( ele )).toMatchSnapshot();
    })
  });

// shallow render GifItem and search in resulting virtual DOM to make sure 
// that mock giphy image prop url has made it to URL of actual image node
describe('check IMG node', () => {
    it('IMG src = mockurl ', () => {
        enzyme.configure({ adapter: new Adapter() });
        const gifitem = enzyme.shallow(ele);
        // query strings for find are called "selectors". There are lots of flavours.
        // see: https://github.com/airbnb/enzyme/blob/master/docs/api/selector.md
        expect(gifitem.contains(<img src="mockurl"/> )).toEqual(true);
    })
});

describe('GifItem calls click handler correctly', () => {
    it('simulate a click on shallow rendering of GifItem', () => {
        enzyme.configure({ adapter: new Adapter() });
        const gifitem = enzyme.shallow( ele );
        // handler has NOT  been called before simulate click 
        expect(mockedCallbackForGifItem.mock.calls.length).toEqual(0);

        gifitem.find('[className="gif-item"]').simulate("click"); 

        // The first argument of the first call to the handler should be mocked GIFObject
        expect(mockedCallbackForGifItem.mock.calls.length).toEqual(1);
        expect(mockedCallbackForGifItem.mock.calls[0][0]).toBe(mockGiphyGifObject); 
    })
});

