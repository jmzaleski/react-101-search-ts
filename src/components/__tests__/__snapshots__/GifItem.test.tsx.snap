// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GiphItem /> renders 1`] = `
<div
  className="gif-item"
  onClick={[Function]}
>
  <img
    alt={undefined}
    src="mockurl"
  />
</div>
`;
